frontend/lib/identification/otp_generator.dart

```diff
@@ -9,11 +12,28 @@ class OTPGenerator {
   OTPGenerator(this._base32TotpSecret, this._otpIntervalSeconds,
       this._otpLength, this._algorithm);
 
-  int generateOTP() {
-    return int.parse(OTP.generateTOTPCodeString(
-        _base32TotpSecret, DateTime.now().millisecondsSinceEpoch,
-        algorithm: _algorithm,
-        length: _otpLength,
-        interval: _otpIntervalSeconds));
+  OTPCode generateOTP([VoidCallback onTimeout]) {
+    final time = DateTime.now().millisecondsSinceEpoch;
+    final validUntilMilliSeconds =
+        ((((time ~/ 1000).round()) ~/ _otpIntervalSeconds).floor() + 1) *
+            _otpIntervalSeconds *
+            1000;
```

Comment:
~/ is the [truncate operator](https://api.flutter.dev/flutter/dart-core/double/operator_truncate_divide.html), so I think, the following should suffice
```suggestion
    final intervalMilliSeconds = _otpIntervalSeconds * 1000;
    final validUntilMilliSeconds = (time ~/ intervalMilliseconds + 1) * intervalMilliSeconds;
```